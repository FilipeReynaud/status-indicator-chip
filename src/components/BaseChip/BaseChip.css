:root {
	--animation-color: #fff;
}

.status-indicator-table {
	min-width: 50px;
	min-height: 25px;
	border-radius: 50px;
	position: relative;
	padding: 0 7px 0 5px;
}

.status-indicator-circle {
	width: 10px;
	height: 10px;
	border-radius: 100%;
}

.status-indicator-label {
	display: block;
	height: 100%;
	font-family: "Roboto", sans-serif;
	font-weight: bold;
}

td:nth-child(2) {
	padding-left: 5px;
	vertical-align: middle;
}

@keyframes blink {
	from {
		background-color: var(--animation-color);
	}

	to {
		background-color: transparent;
	}
}
