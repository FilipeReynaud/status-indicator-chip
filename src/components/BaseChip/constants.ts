const CLASS_NAMES = {
	statusIndicatorTable: "status-indicator-table",
	statusIndicatorCircle: "status-indicator-circle",
	statusIndicatorLabel: "status-indicator-label",
};

const DATA_TEST_IDS = {
	...CLASS_NAMES,
};

const FALLBACK_CSS_VALUES = {
	color: "#0eaf00",
	backgroundColor: "#ddecd1",
};

const LABEL_SIZE = {
	sm: "8px",
	md: "11px",
	lg: "16px",
};

const DEFAULT_VALUES = {
	label: "Active",
	size: "md",
};

export {
	DATA_TEST_IDS,
	CLASS_NAMES,
	FALLBACK_CSS_VALUES,
	LABEL_SIZE,
	DEFAULT_VALUES,
};
