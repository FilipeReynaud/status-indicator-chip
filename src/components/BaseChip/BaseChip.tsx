import React, { CSSProperties } from "react";

// Styles
import "./BaseChip.css";

// Constants
import {
	CLASS_NAMES,
	DEFAULT_VALUES,
	FALLBACK_CSS_VALUES,
	LABEL_SIZE,
} from "./constants";

// Types
interface BaseChipCSS extends CSSProperties {
	"--animation-color": string;
}

interface BaseChipProps {
	fontColor: string;
	backgroundColor: string;
	label: string;
	withBlinkAnimation?: boolean;
	size?: "sm" | "md" | "lg";
}

const appendStylesTo = (className: string, styles: BaseChipProps) => {
	if (className === CLASS_NAMES.statusIndicatorTable) {
		return {
			backgroundColor:
				styles.backgroundColor ?? FALLBACK_CSS_VALUES.backgroundColor,
		};
	} else if (className === CLASS_NAMES.statusIndicatorCircle) {
		return {
			backgroundColor: styles.fontColor ?? FALLBACK_CSS_VALUES.color,
			"--animation-color": styles.withBlinkAnimation
				? styles.fontColor ?? FALLBACK_CSS_VALUES.color
				: "",
			animationName: styles.withBlinkAnimation ? "blink" : "",
			animationDuration: "1s",
			animationIterationCount: "infinite",
		} as BaseChipCSS;
	} else if (className === CLASS_NAMES.statusIndicatorLabel) {
		return {
			color: styles.fontColor,
			fontSize: styles.size
				? LABEL_SIZE[styles.size]
				: LABEL_SIZE[DEFAULT_VALUES.size as keyof typeof LABEL_SIZE],
		};
	}

	return {};
};

const BaseChip = (props: BaseChipProps) => (
	<table
		className={CLASS_NAMES.statusIndicatorTable}
		style={appendStylesTo(CLASS_NAMES.statusIndicatorTable, props)}
	>
		<tbody>
			<tr>
				<td>
					<div
						className={CLASS_NAMES.statusIndicatorCircle}
						style={appendStylesTo(
							CLASS_NAMES.statusIndicatorCircle,
							props
						)}
					></div>
				</td>
				<td>
					<span
						className={CLASS_NAMES.statusIndicatorLabel}
						style={appendStylesTo(
							CLASS_NAMES.statusIndicatorLabel,
							props
						)}
					>
						{props.label ?? DEFAULT_VALUES.label}
					</span>
				</td>
			</tr>
		</tbody>
	</table>
);

export default BaseChip;
