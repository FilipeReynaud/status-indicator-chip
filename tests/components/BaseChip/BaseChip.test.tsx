import React from "react";
import { render } from "@testing-library/react";

// Components
import BaseChip from "../../../src/components/BaseChip";

// Types
import { BaseChipProps } from "../../../src/components/BaseChip/BaseChip";

describe("BaseChip", () => {
	const defaultTestProps = {
		fontColor: "blue",
		backgroundColor: "lightblue",
		label: "Test",
		withBlinkAnimation: true,
		size: "md",
	} as BaseChipProps;

	it("should render properly", () => {
		// Act
		const { container } = render(<BaseChip {...defaultTestProps} />);

		// Assert
		expect(container).toMatchSnapshot();
	});
});
