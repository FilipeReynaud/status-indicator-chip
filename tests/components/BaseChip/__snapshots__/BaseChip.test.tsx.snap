// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseChip should render properly 1`] = `
<div>
  <table
    class="status-indicator-table"
    style="background-color: lightblue;"
  >
    <tbody>
      <tr>
        <td>
          <div
            class="status-indicator-circle"
            style="background-color: blue; --animation-color: blue; animation-name: blink; animation-duration: 1s; animation-iteration-count: infinite;"
          />
        </td>
        <td>
          <span
            class="status-indicator-label"
            style="color: blue; font-size: 11px;"
          >
            Test
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
